packages/server-wallet/src/wallet/wallet.ts

```diff
@@ -948,68 +944,32 @@ export class SingleThreadedWallet
    * @param channels channels touched by the caller
    * @param response WalletResponse that is modified in place while cranking objectives
    */
-  private async takeActions(channels: Bytes32[], response: WalletResponse): Promise<void> {
-    let needToCrank = true;
-    while (needToCrank) {
-      await this.crankUntilIdle(channels, response);
-      needToCrank = await this.processLedgerQueue(channels, response);
-    }
+  private async takeActions(channelIds: Bytes32[], response: WalletResponse): Promise<void> {
+    // 1. we're passed a set of channelIds that have changed
+    // 2. we crank any objectives that involve changed channels, which could create new ledger requests
+    // 3. we crank any ledgers with new requests, which could update those requests
+    // 4. if any requests changed, we treat the corresponding channels as changed and go to step 2
+
+    while (channelIds.length > 0) {
+      const objectiveIds = await this.store.getApprovedObjectiveIds(channelIds);
+      for (const objectiveId of objectiveIds) {
+        await this.objectiveManager.crank(objectiveId, response);
+      }
 
-    response.createdObjectives.map(o => this.emit('objectiveStarted', o));
-    response.succeededObjectives.map(o => this.emit('objectiveSucceeded', o));
-  }
+      const ledgersFromRequests = await this.store.getLedgersWithNewRequestsIds();
```

Comment:
This is a follow up from a previous question. These are the ledger requests with an empty `lastSeenAgreedState`. What would be the impact on this code if `getLedgersWithNewRequestsIds` returned `queued` requests?

Reply:
I am trying to think through the following scenario. Let's say that A and B would like to ledger fund channels c and d using ledger channel L. L's initial outcome is [A, B]. Let's walk through the following sequence of events:
1. A calls `createChannel`  for c, then A calls `createChannel` for d.
2. B calls `joinChannel` for c, then B calls `joinChannel` for d.
3. B now has ledger requests for c and d.
4. A receives B's prefund signature for c. B's prefund signature on d gets delayed in transit.
5. A sends a ledger state Proposal that only includes c: [A, B, c].
6. B signs the Proposal [A, B, c] and sets the `lastSeenState` for d.
7. B's prefund signature on d finally arrives to A.
8. A receives B's signature on Proposal [A, B, c]. A now sends a second Proposal that includes d: [A, B, c, d].

I can't seem to convince myself that B's `takeActions` will process the second Proposal. I am probably missing something...

Reply:
Maybe I am starting to understand. When B receives the Proposal of [A, B, c, d], `takeActions` should be called with `channelIds` parameter of `L`. So ` await this.store.filterChannelIdsByIsLedger(channelIds)` should result in the proposal getting processed.

Reply:
In the case above, the ledger would be included in `ledgersFromChannels` (the line below this one), and so would get processed by b. The general logic is that there are two cases where you need to crank a ledger:

1. When that ledger channel itself has received an update
2. When an objective has just added a new request to the queue

If the case where a ledger has (a) already seen all of its requests, and (b) hasn't changed, you don't need to crank again, as you'll just get the same result. So selecting `queued` requests here would lead to redundant work.

I guess there's an argument for adding a `new` status to the ledger request, to make the case of "not yet processed by a ledger crank" more explicit.

Reply:
Thanks for the explanation!